core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1899,6 +1899,54 @@ object Stream {
     def foldMonoid(implicit O: Monoid[O]): Stream[F, O] =
       self.fold(O.empty)(O.combine)
 
+    def groupWithin(n: Int, d: FiniteDuration)(
+        implicit F: Effect[F],
+        timer: Timer[F],
+        ec: ExecutionContext): Stream[F, Segment[O, Unit]] = {
+      type Tick = Int
+
+      Stream.eval(async.unboundedQueue[F, Option[Either[Tick, O]]]).flatMap { q =>
+        def startTimeout(i: Tick): Stream[F, Nothing] =
+          Stream.eval {
+            async.fork {
+              timer.sleep(d) *> q.enqueue1(i.asLeft.some)
+            }
+          }.drain
+
+        def producer = self.map(Right(_)).noneTerminate.to(q.enqueue)
+
+        // note: if empty sequences are emitted instead, a `startTimeout` before `go` is needed
+        def emitNonEmpty(c: Catenable[Segment[O, Unit]]): Stream[F, Segment[O, Unit]] =
+          if (c.nonEmpty) Stream.emit(Segment.catenated(c))
+          else Stream.empty
+
+        def go(acc: Catenable[Segment[O, Unit]],
+               elems: Int,
+               currentTimeout: Tick): Stream[F, Segment[O, Unit]] =
+          Stream.eval(q.dequeue1).flatMap {
+            case None => emitNonEmpty(acc)
+            case Some(e) =>
+              e match {
+                case Left(t) if t > currentTimeout => throw new Exception
```

Comment:
Do you want to throw instead of `Stream.raiseError`?

What does this case mean semantically? How is this branch hit?

Reply:
that's a wip line. I want an `error` eventually because that branch can't be hit I don't think. Sloppy writing to get this in tonight :)